api4/post_test.go

```diff
@@ -2527,3 +2527,84 @@ func TestGetFileInfosForPost(t *testing.T) {
 	_, resp = th.SystemAdminClient.GetFileInfosForPost(th.BasicPost.Id, "")
 	CheckNoError(t, resp)
 }
+
+func TestSetChannelUnread(t *testing.T) {
+	th := Setup().InitBasic()
+	defer th.TearDown()
+
+	u1 := th.BasicUser
+	u2 := th.BasicUser2
+	s2, _ := th.App.GetSession(th.Client.AuthToken)
+	th.Client.Login(u1.Email, u1.Password)
+	c1 := th.BasicChannel
+	c1toc2 := &model.ChannelView{ChannelId: th.BasicChannel2.Id, PrevChannelId: c1.Id}
+	now := utils.MillisFromTime(time.Now())
+	p1 := th.CreateMessagePostNoClient(c1, "AAA", now)
+	require.NotNil(t, p1)
+	p2 := th.CreateMessagePostNoClient(c1, "BBB", now+10)
+	require.NotNil(t, p2)
+	p3 := th.CreateMessagePostNoClient(c1, "CCC", now+20)
+	require.NotNil(t, p3)
+
+	pp1 := th.CreateMessagePostNoClient(th.BasicPrivateChannel, "Sssh!", now)
+	pp2 := th.CreateMessagePostNoClient(th.BasicPrivateChannel, "You Sssh!", now+10)
+	require.NotNil(t, pp1)
+	require.NotNil(t, pp2)
+
+	t.Run("Can't unread if user is not logged in", func(t *testing.T) {
+		th.Client.Logout()
+		response := th.Client.SetPostUnread(u1.Id, p2.Id)
+		checkHTTPStatus(t, response, http.StatusUnauthorized, true)
+	})
+
+	th.Client.Login(u1.Email, u1.Password)
+
+	t.Run("Ensure that post have been read", func(t *testing.T) {
+		unread, err := th.App.GetChannelUnread(c1.Id, u1.Id)
+		require.Nil(t, err)
+		require.Equal(t, int64(4), unread.MsgCount)
+		unread, err = th.App.GetChannelUnread(c1.Id, u2.Id)
+		require.Nil(t, err)
+		require.Equal(t, int64(4), unread.MsgCount)
+		_, err = th.App.ViewChannel(c1toc2, u2.Id, s2.Id)
+		require.Nil(t, err)
+		unread, err = th.App.GetChannelUnread(c1.Id, u2.Id)
+		require.Nil(t, err)
+		require.Equal(t, int64(0), unread.MsgCount)
+	})
+
+	t.Run("Unread last one", func(t *testing.T) {
+		r := th.Client.SetPostUnread(u1.Id, p2.Id)
+		checkHTTPStatus(t, r, 200, false)
+		unread, err := th.App.GetChannelUnread(c1.Id, u1.Id)
+		require.Nil(t, err)
+		assert.Equal(t, int64(1), unread.MsgCount)
+	})
+
```

Comment:
Maybe add a test case for when you set the second or third post as unread to make sure that the `MsgCount` increases correctly?

Reply:
adding it on the app test instead